lambda-buffers-compiler/test/Test/TypeClassCheck.hs

```diff
@@ -70,3 +101,219 @@ nocycles =
   , mkclass "Monad" ["Applicative"]
   , mkclass "Traversable" ["Foldable", "Functor"]
   ]
+
+---- Solver tests
+
+solveTests :: TestTree
+solveTests =
+  testGroup
+    "Solver tests"
+    [ testCase "1: C [Maybe Int] (completeRules)" $
+        solveTest1 @?= []
+    , testCase "2: D [Maybe Int] (partialRules)" $
+        solveTest2 @?= [dInt]
+    , testCase "3: D [Maybe Int] (complete D, partial C)" $
+        solveTest3 @?= [cInt]
+    , testCase "4: C [[[Bool]]] (completeRules)" $
+        solveTest4 @?= []
+    , testCase "5: C (Either (Either Int Bool) (Either Bool Int)) (completeRules)" $
+        solveTest5 @?= []
+    , testCase "6: C (Either l x) (completeRules)" $
+        solveTest6 @?= [cL, cX]
+    , testCase "7: Sum test (completeRules)" $
+        solveTest7 @?= []
+    , testCase "8: Sum test (partialRules)" $
+        solveTest8 @?= [cBool, cInt]
+    , testCase "9: Rec test (completeRules)" $
+        solveTest9 @?= []
+    , testCase "10: Rec test (partialRules)" $
+        solveTest10 @?= [cBool, cInt]
+    , testCase "11: Prod test (completeRules)" $
+        solveTest11 @?= []
+    , testCase "12: Prod test (partialRules)" $
+        solveTest12 @?= [cBool, cInt]
+    , testCase "13: Dec test (completeRules)" $
+        solveTest13 @?= []
+    , testCase "14: Dec test (partialRules)" $
+        solveTest14 @?= [cBool, cInt]
+    ]
+  where
+    cInt = C _c Int
+    cBool = C _c Bool
+    cL = C _c _l
+    cX = C _c _x
+    dInt = C _d Int
+
+-- hardcoded variables, easier to read than (VarP "blah") everywhere
+_x, _xs, _vars, _name, _l, _body :: Pat
+_x = VarP "x"
+_xs = VarP "xs"
+_vars = VarP "vars"
+_name = VarP "name"
+_l = VarP "l"
+_body = VarP "body"
+
+-- TODO(@gnumonik): Import this from somewhere else when that somewhere else module exists
+mkStructuralRules :: Class -> [Rule]
+mkStructuralRules c =
+  [ C c Nil :<= []
+  , C c (_x :* _xs) :<= [C c _x, C c _xs]
+  , C c (_l := _x) :<= [C c _x]
+  , C c (RecP _xs) :<= [C c _xs]
+  , C c (ProdP _xs) :<= [C c _xs]
+  , C c (SumP _xs) :<= [C c _xs]
+  , C c (DecP _name _vars _body) :<= [C c _body]
+  ]
+
+pattern (:@) :: Pat -> Pat -> Pat
+pattern (:@) p1 p2 = AppP p1 p2
+
+pattern LocalRef :: Text -> Pat
+pattern LocalRef nm = RefP Nil (Name nm)
+
+pattern Maybe :: Pat -> Pat
+pattern Maybe p1 = LocalRef "Maybe" :@ p1
+
+pattern List :: Pat -> Pat
+pattern List p = LocalRef "List" :@ p
+
+pattern Either :: Pat -> Pat -> Pat
+pattern Either l r = (LocalRef "Either" :@ l) :@ r
+
+pattern Int :: Pat
+pattern Int = LocalRef "Int"
+
+pattern Bool :: Pat
+pattern Bool = LocalRef "Bool"
+
+pattern NoConstraints :: Class -> Pat -> Rule
+pattern NoConstraints c p = C c p :<= []
+
+pattern Labeled :: Text -> Pat -> Pat
+pattern Labeled nm p = Name nm := p
+
+userRules1 :: Class -> [Rule]
+userRules1 c =
+  [ NoConstraints c Int
+  , NoConstraints c Bool
+  , C c (Maybe _x) :<= [C c _x]
+  , C c (Either _l _x) :<= [C c _l, C c _x]
+  , C c (List _x) :<= [C c _x]
```

Comment:
I don't think this does what we need it to do...

Consider...

```haskell

-- Int and String don't have an Eq

data Foo a = MkFoo Int String (Maybe a)

instance Eq a => Eq (Maybe a)

instance Eq a => Eq (Foo a)
```

Reply:
What happens in your implementation?

Reply:
Which rules fire?

Reply:
Are we trying to derive both of those instances or are we assuming them as rules? 

If we try to derive them, we have to "pretend" that we have `Eq a` in scope (i.e. we assume it). But - and I'm only just realizing this now - what we specifically need to assume is that `Eq a` holds for the _type variable_ `a` and not the _unification variable_ `a`. So, as I hinted at above, I need to distinguish `VarP` from something like `TyVarP` (where `TyVarP "a"` is a _literal pattern_ or ground term if you like). 

This is the correct thing to assume for codegen because when writing the methods for an  `instance Eq a => Eq (Foo a)` , the compiler assumes that `a` is an `Eq` instance. (Once you try to write these you'll realize that all the codegen for the instances looks the same, we're just trying to figure out whether it's OK to generate the code or whether we should report an error and stop prior to code generation) 

`solve` isn't meant to be the sum total of the derivation machinery, and what exactly is in scope depends upon which of the instances we're trying to derive. Let's pretend we're trying to derive `Eq a => Eq (Foo a)` b/c that's more interesting: 
 
Let's convert to a pattern 

```
foo = DecP (Name "Foo) (Name "a" :* Nil)  $ SumP (Name "MkFoo" := ProdP ( Int :* String :* Maybe (TyVarP "a") :* Nil)) :* Nil   
```

Nothing fires w/o the structural rules (obviously) but assume we have those in scope, instances for `Int/String` in scope, and `Eq a` (where `a` is a literal `TyVarP "a"`, so really `Eq (TyVarP "a")` ) as a consequence of the above assumption procedure.  We also have `Eq a => Eq (Maybe a)` in scope here.  So our scope is something like: 

``` 
-- structural 
C Eq Nil :<= []
C Eq (_x :* _xs) :<= [C Eq _x, C Eq _xs]
C Eq (_l := _x) :<= [C Eq _x]
C Eq (RecP _xs) :<= [C Eq _xs]
C Eq (ProdP _xs) :<= [C Eq _xs]
C Eq (SumP _xs) :<= [C Eq _xs]
C Eq (DecP _name _vars _body) :<= [C Eq _body]
-- for the sake of the example 
C Eq Int 
C Eq String 
-- b/c we assume all other instances in the same module are in scope 
C Eq (Maybe _x) :<= [C Eq _x]
-- added assumption for constrained instances explained above 
C Eq (TyVarP "a")
```
Reduction:
```
-- via structural rule for DecP 
C (SumP (Name "mkFoo" := ProdP ( Int :* String :* Maybe (TyVarP "a") :* Nil)) :* Nil) 

-- via structural rules for SumP 
C ((Name "mkFoo" := ProdP ( Int :* String :* Maybe (TyVarP "a") :* Nil)) :* Nil)

-- via structural rule for Nil 
C (Name "mkFoo" := ProdP ( Int :* String :* Maybe (TyVarP "a") :* Nil)) 

-- via structural rule for (:=) 
C (ProdP ( Int :* String :* Maybe (TyVarP "a") :* Nil)

-- via structural rule for ProdP 
C ( Int :* String :* Maybe (TyVarP "a") :* Nil)

-- via structural rule for (:*) and (Eq Int)   - I'm starting at the head but we could do it the other way 
C (String :* Maybe (TyVarP "a") :* Nil)

-- via structural rule for (:*) and (Eq String) 
C (Maybe (TyVarP "a") :* Nil) 

-- via structural rules for :* and Nil 
C (Maybe (TyVarP "a") 

-- via rule for Maybe 
C (TyVarP "a") 

-- via assumption of (TyVarP "a") 
()
```

Reply:
But do note that `solve` is just supposed to be the function you call _once you've already established the scope (including any additional assumptions)_. The process of establishing those assumptions is a separate concern.

Reply:
@bladyjoker it's important that this is right, you can skim most of the other responses but pls read this one closely

Reply:
Let me make a better example!

1. Int has Eq
2. String doesn't have Eq

Let's consider both use cases, namely `derive` and `instance`:

```haskell
opaque Int
opaque String
data Foo a = MkFoo Int String (Maybe a)

instance Eq Int
instance Eq a => Eq (Maybe a)
instance Eq a => Eq (Foo a)
```
vs.

```haskell
opaque Int
opaque String
data Foo a = MkFoo Int String (Maybe a)

derive Eq Int
derive Eq (Maybe a)
derive Eq (Foo a)
```
In the `instance` case, this is a 'user supplied clause` and we just put it in amongst other rules:

```haskell
-- structural 
C Eq Nil :<= []
C Eq (_x :* _xs) :<= [C Eq _x, C Eq _xs]
C Eq (_l := _x) :<= [C Eq _x]
C Eq (RecP _xs) :<= [C Eq _xs]
C Eq (ProdP _xs) :<= [C Eq _xs]
C Eq (SumP _xs) :<= [C Eq _xs]
C Eq (DecP _name _vars _body) :<= [C Eq _body]

-- user specified
C Eq Int :<= []
C Eq (Maybe _x) :<= [C Eq _x]
C Eq (Foo _x) :<= [C Eq _x]
```

When I ask `Eq (Foo Int)` what happens? Given this rule set:

1. rule `C Eq (Foo _x) :<= [C Eq _x]` fires
2. rule `C Eq Int :<= []` fires
3. done --- which is correct as far as the Compiler is concerned right?

Unless I'm not seeing why necessarily `C Eq (DecP _name _vars _body) :<= [C Eq _body]` would fire??? The user can't really specify the rules we construct with `derive`.

However, in the `derive` case:

```haskell
-- structural 
C Eq Nil :<= []
C Eq (_x :* _xs) :<= [C Eq _x, C Eq _xs]
C Eq (_l := _x) :<= [C Eq _x]
C Eq (RecP _xs) :<= [C Eq _xs]
C Eq (ProdP _xs) :<= [C Eq _xs]
C Eq (SumP _xs) :<= [C Eq _xs]
C Eq (DecP _name _vars _body) :<= [C Eq _body]

-- specified but derived rules
C Eq Int :<= []
C Eq (Maybe _x) :<= [C Eq (DecP _name _vars _body)] % This should be C Eq <Maybe applied to _x so Just _ | Nothing>
C Eq (Foo _x) :<= [C Eq (DecP _name _vars _body)] % <Foo applied to _x so MkFoo Int String (Maybe _x)>
```

Now, when I ask `Eq (Foo Int)` what happens? Given this rule set:
1. `C Eq (Foo _x) :<= [C Eq (DecP _name _vars _body)] ` fires
2. `C Eq (DecP Foo _vars "MkFoo Int String (Maybe _x)") fires
3. Sum/Product rules kick in until
4. `C Eq String` is missing!!!
5. done

Reply:
> % IMPORTANT!!!: Just realized that this is unnecessary to be part of the Solver, could be, but rather we can construct these rules of each of the derive goals, which is what the user expects! AND WE DON'T NEED THE EXTRA Apply and TyDef 'utility' constraints

oh jfc i didn't scroll to the right and spent like an hour and a half today trying to figure out how to explain that to you lol

Reply:
lol sorry about that :D